File: algorithms/Strings/Medium/FractionToRecurringDecimal/VSDFractionToRecurringDecimal.py



def fractionToDecimal(numerator,denominator):

    if numerator%denominator==0:
        return str(numerator//denominator)
    
    sign="" if numerator * denominator >=0 else "-"

    numerator,denominator=abs(numerator),abs(denominator)

    res=sign+str(numerator//denominator)+"."

    numerator%=denominator
    i,part=0,""
    m={numerator:i}
    while numerator%denominator:
        numerator*=10
        i+=1
        rem=numerator%denominator
        part+=str(numerator//denominator)
        if rem in m:
            part=part[:m[rem]]+"("+part[m[rem]:]+")"
            return res+part
        m[rem]=i
        numerator=rem 
    return res+part 

if __name__=="__main__":
    n=1
    d=2 
    print(fractionToDecimal(n,d))

File: algorithms/Strings/Medium/FractionToRecurringDecimal/VSDFractionToRecurringDecimal.cpp

#include <bits/stdc++.h>

std::string tostring(long long n)
{
    if (n == 0)
        return "0";
    std::string ans;
    while (n > 0)
    {
        std::string rem;
        rem += ((n % 10) - '0');
        std::cout<<rem<<std::endl;
        ans = rem + ans;
        n /= 10;
    }
    return ans;
}

std::string fractionToDecimal(int numerator, int denominator)
{

    if (numerator == 0)
        return "0";
    std::string ans="";
    if ((numerator < 0 ^ denominator < 0))
        ans.push_back('-');
    long long num = abs(numerator);
    long long denum = abs(denominator);
    ans = ans+tostring(num/denum);
    num%=denum;
    if(num==0) return ans;

    ans.push_back('.');
    std::map<long long,int> map;
    map[num]=ans.length();

    while(num != 0){
        num *= 10;
        
        ans.push_back('0'+(num / denum));
        num %=denum;
        if(map.find(num) != map.end()){
            int index=map[num];
            
            ans.insert(ans.begin()+index,'(');
            ans.push_back(')');
            return ans;
        }
        else{
            map[num]=ans.length();

        }
    }
    return ans;
}

int main()
{
    int n = 1;
    int d = 2;
    std::cout << fractionToDecimal(n, d);
}